fms2_io/fms_netcdf_domain_io.F90

```diff
@@ -546,44 +546,55 @@ subroutine save_domain_restart(fileobj, unlim_dim_level)
+    if (associated(fileobj%restart_vars(i)%data2d)) then
       chksum = compute_global_checksum(fileobj, fileobj%restart_vars(i)%varname, &
                                        fileobj%restart_vars(i)%data2d, is_decomposed)
       if (is_decomposed) then
         call register_variable_attribute(fileobj, fileobj%restart_vars(i)%varname, &
                                          "checksum", chksum)
       endif
     elseif (associated(fileobj%restart_vars(i)%data3d)) then
-      call domain_write_3d(fileobj, fileobj%restart_vars(i)%varname, &
-                           fileobj%restart_vars(i)%data3d, unlim_dim_level=unlim_dim_level)
       chksum = compute_global_checksum(fileobj, fileobj%restart_vars(i)%varname, &
                                        fileobj%restart_vars(i)%data3d, is_decomposed)
       if (is_decomposed) then
         call register_variable_attribute(fileobj, fileobj%restart_vars(i)%varname, &
                                          "checksum", chksum)
       endif
     elseif (associated(fileobj%restart_vars(i)%data4d)) then
-      call domain_write_4d(fileobj, fileobj%restart_vars(i)%varname, &
-                           fileobj%restart_vars(i)%data4d, unlim_dim_level=unlim_dim_level)
       chksum = compute_global_checksum(fileobj, fileobj%restart_vars(i)%varname, &
                                        fileobj%restart_vars(i)%data4d, is_decomposed)
       if (is_decomposed) then
         call register_variable_attribute(fileobj, fileobj%restart_vars(i)%varname, &
                                          "checksum", chksum)
       endif
+    endif
+  enddo
+
+! Write out all of the data
+  do i = 1, fileobj%num_restart_vars
+    if (associated(fileobj%restart_vars(i)%data0d)) then
+      call domain_write_0d(fileobj, fileobj%restart_vars(i)%varname, &
+                           fileobj%restart_vars(i)%data0d, unlim_dim_level=unlim_dim_level)
+    elseif (associated(fileobj%restart_vars(i)%data1d)) then
+      call domain_write_1d(fileobj, fileobj%restart_vars(i)%varname, &
+                           fileobj%restart_vars(i)%data1d, unlim_dim_level=unlim_dim_level)
+    elseif (associated(fileobj%restart_vars(i)%data2d)) then
+      call domain_write_2d(fileobj, fileobj%restart_vars(i)%varname, &
+                           fileobj%restart_vars(i)%data2d, unlim_dim_level=unlim_dim_level)
+    elseif (associated(fileobj%restart_vars(i)%data3d)) then
+      call domain_write_3d(fileobj, fileobj%restart_vars(i)%varname, &
+                           fileobj%restart_vars(i)%data3d, unlim_dim_level=unlim_dim_level)
+    elseif (associated(fileobj%restart_vars(i)%data4d)) then
+      call domain_write_4d(fileobj, fileobj%restart_vars(i)%varname, &
+                           fileobj%restart_vars(i)%data4d, unlim_dim_level=unlim_dim_level)
     else
-      call error("this branch should not be reached.")
+      call error("save_domain_restart: this branch should not be reached.")
```

Comment:
This error message is still not descriptive of the problem.  If there is a fatal, then I know that I shouldn't have gotten here.  The user will be puzzled as to why they got here or how to fix it.